modules/navier_stokes/src/actions/NSFVAction.C

```diff
@@ -0,0 +1,2289 @@
+  {
+    auto params = _factory.getValidParams("MooseVariableFVReal");
+    params.set<std::vector<SubdomainName>>("block") = _blocks;
+
+    _problem->addAuxVariable("MooseVariableFVReal", NS::mixing_length, params);
+  }
+
+  if (_has_energy_equation)
+  {
+    auto params = _factory.getValidParams("INSFVEnergyVariable");
+    params.set<std::vector<SubdomainName>>("block") = _blocks;
+    params.set<std::vector<Real>>("scaling") = {_energy_scaling};
+    params.set<MooseEnum>("face_interp_method") = _energy_face_interpolation;
+    params.set<bool>("two_term_boundary_expansion") = _energy_two_term_bc_expansion;
+
+    _problem->addVariable("INSFVEnergyVariable", NS::T_fluid, params);
+  }
+
+  if (_passive_scalar_names.size())
+  {
+    auto params = _factory.getValidParams("MooseVariableFVReal");
+    params.set<std::vector<SubdomainName>>("block") = _blocks;
+    params.set<std::vector<Real>>("scaling") = {_passive_scalar_scaling};
+    params.set<MooseEnum>("face_interp_method") = _passive_scalar_face_interpolation;
+    params.set<bool>("two_term_boundary_expansion") = _passive_scalar_two_term_bc_expansion;
+
+    for (unsigned int name_i = 0; name_i < _passive_scalar_names.size(); ++name_i)
+    {
+      bool create_me = true;
+      if (_create_scalar_variable.size())
+        if (!_create_scalar_variable[name_i])
+          create_me = false;
+
+      if (create_me)
+        _problem->addVariable("MooseVariableFVReal", _passive_scalar_names[name_i], params);
+    }
+  }
+}
+
+void
+NSFVAction::addRhieChowUserObjects()
+{
+  const std::string u_names[3] = {"u", "v", "w"};
+  if (_porous_medium_treatment)
+  {
+    auto params = _factory.getValidParams("PINSFVRhieChowInterpolator");
+    params.set<std::vector<SubdomainName>>("block") = _blocks;
+    for (unsigned int d = 0; d < _dim; ++d)
+      params.set<VariableName>(u_names[d]) = NS::superficial_velocity_vector[d];
+
+    params.set<VariableName>("pressure") = NS::pressure;
+    params.set<MooseFunctorName>(NS::porosity) = _porosity_name;
+    unsigned short smoothing_layers = isParamValid("porosity_smoothing_layers")
+                                          ? getParam<unsigned short>("porosity_smoothing_layers")
+                                          : 0;
+    params.set<unsigned short>("smoothing_layers") = smoothing_layers;
+
+    _problem->addUserObject("PINSFVRhieChowInterpolator", "pins_rhie_chow_interpolator", params);
+  }
+  else
+  {
+    auto params = _factory.getValidParams("INSFVRhieChowInterpolator");
+    params.set<std::vector<SubdomainName>>("block") = _blocks;
+    for (unsigned int d = 0; d < _dim; ++d)
+      params.set<VariableName>(u_names[d]) = NS::velocity_vector[d];
+
+    params.set<VariableName>("pressure") = NS::pressure;
+
+    _problem->addUserObject("INSFVRhieChowInterpolator", "ins_rhie_chow_interpolator", params);
+  }
+}
+
+void
+NSFVAction::addINSInitialConditions()
+{
+  InputParameters params = _factory.getValidParams("ConstantIC");
```

Comment:
ah I guess we dont have a FunctorIC yet..
want to make one or should I do it?

Reply:
Let's do it after this guy is in.